### PR TITLE
Moved fits import above pillow and added unit test of get_reader

### LINKED ISSUE #636: "cannot find loader for this FITS file" error
Hi, I was looking at the `imageio` module and I see that the FITS format is supposedly supported ([here](https://imageio.readthedocs.io/en/stable/formats.html)). But, I get an error when I try to use `imageio.get_reader()` on a file without the `format` keyword set to fits. So, I guess there's some mismatch in how `imagio` is trying to automatically determine the file type.

For a minimum reproducible example, you can stand up this environment:
```
conda create -n testio -c conda-forge scikit-image astropy imageio-ffmpeg pims python=3.7
```

After activating the environment, in Python generate a fake image and try reading it in:
```
import numpy as np
from astropy.io import fits
from astropy.modeling.models import Sersic2D
import imageio
xx, yy = np.meshgrid(np.arange(512),np.arange(512))
mod = Sersic2D(amplitude = 1, r_eff = 25, n=4, x_0=50,y_0=50, ellip=.5, theta=-1)
img = mod(xx, yy)
phdu = fits.PrimaryHDU()
ihdu = fits.ImageHDU(img)
hdul = fits.HDUList([phdu,ihdu])
hdul.writeto('test.fits')
im = imageio.get_reader('test.fits')
```
The last line fails for me with the following error:
```
Traceback (most recent call last):
  File "/local/data/home/[user]/.conda/envs/testio/lib/python3.7/site-packages/imageio/plugins/pillow.py", line 669, in pil_try_read
    im.getdata()[0]
  File "/local/data/home/[user]/.conda/envs/testio/lib/python3.7/site-packages/PIL/Image.py", line 1292, in getdata
    self.load()
  File "/local/data/home/[user]/.conda/envs/testio/lib/python3.7/site-packages/PIL/ImageFile.py", line 327, in load
    raise OSError(f"cannot find loader for this {self.format} file")
OSError: cannot find loader for this FITS file
```

But, I noticed if I give the format like the following line, the `get_reader()` call succeeds:
```
im = imageio.get_reader('test.fits',format='fits')
```

### PR DESCRIPTION
**Pull request by astrophysaxist**
Moved the `import fits` line in `imageio/plugins/__init__.py` to occur before `import pillows` to close #636.  Also added a unit test in `tests/test_fits.py` for `imageio.get_reader()`.

### CODE REVIEW

File: tests/test_fits.py [hunk 1]
```diff
@@ -79,5 +80,31 @@ def test_fits_reading():
     im = imageio.imread(compressed)
     assert im.shape == (2042, 3054)
 
+@pytest.mark.skipif("astropy is None")
+def test_fits_get_reader():
```
Comment (FirefoxMetzger) [inline tests/test_fits.py:84]:
```suggestion
def test_fits_get_reader(tmp_path):
```

We can use the pytest fixture `tmp_path` here to get a temorary directory that will be automatically cleaned up after the test.

File: tests/test_fits.py [hunk 2]
```diff
@@ -79,5 +80,31 @@ def test_fits_reading():
...
+    import numpy as np
+    from astropy.io import fits
+    sigma = 10
+    xx, yy = np.meshgrid(np.arange(512),np.arange(512))
+    z = (1/(2*np.pi*(sigma**2)))*np.exp(-((xx**2)+(yy**2))/(2*(sigma**2)))
+    img = np.log(z)
+    phdu = fits.PrimaryHDU()
+    ihdu = fits.ImageHDU(img)
+    hdul = fits.HDUList([phdu,ihdu])
+    hdul.writeto('test.fits')
```
Comment (FirefoxMetzger) [inline tests/test_fits.py:102]:
```suggestion
    hdul.writeto(tmp_path / 'test.fits')
```

We can use above mentioned tmp_path here to create the file inside the temporary directory.

File: tests/test_fits.py [hunk 3]
```diff
@@ -79,5 +80,31 @@ def test_fits_reading():
...
+    try:
+        im = imageio.get_reader('test.fits')
+        os.remove('test.fits')
+    except ValueError:
+        os.remove('test.fits')
+        pytest.fail('imageio.get_reader failed to find FITS loader')
```
Comment (FirefoxMetzger) [inline tests/test_fits.py:108]:
```suggestion
    im = imageio.get_reader('test.fits')
```

We can refactor the `os.remove` call, because we can use tmp_path (it is cleaner).

If we couldn't refactor `os.remove`, it would be cleaner to write

```python
try:
    pass # code here
except ValueError:
    pass # handle exception
finally:
    # will _always_ be executed, regardless of success or failure
    os.remove(...)
```

If a test produces an unhandled exception, pytest will mark the test as failed automatically, hence we do not have to specifically tell pytest to fail manually.

File: tests/test_fits.py [hunk 4]
```diff
@@ -79,5 +80,31 @@ def test_fits_reading():
     im = imageio.imread(compressed)
     assert im.shape == (2042, 3054)
 
+@pytest.mark.skipif("astropy is None")
+def test_fits_get_reader():
+    """Test reading fits with get_reader method"""
```
Comment (FirefoxMetzger) [inline tests/test_fits.py:85]: Could you mention the GitHub issue (#636 ) that resulted in this regression test in the above comment? This way, it will be clear that it is a regression test, and we also have a reference to its origin,

File: tests/test_fits.py [hunk 5]
```diff
@@ -79,5 +80,31 @@ def test_fits_reading():
...
+    
+    #Set precedence as normal
+    teardown_module()
+    
+    if IS_PYPY:
+        return  # no support for fits format :(
```
Comment (FirefoxMetzger) [inline tests/test_fits.py:91]: I am aware that we do the same in other parts of the codebase, but it is actually an anti-pattern; likely from legacy code before pytest had a cleaner way to do things. We have to refactor this sooner or later ;)

Could you remove this and instead add the decorator `@pytest.mark.skipif("IS_PYPY", reason="pypy doesn't support astropy.fits.")`?

File: tests/test_fits.py [hunk 6]
```diff
@@ -79,5 +80,31 @@ def test_fits_reading():
     im = imageio.imread(compressed)
     assert im.shape == (2042, 3054)
 
+@pytest.mark.skipif("astropy is None")
+def test_fits_get_reader():
+    """Test reading fits with get_reader method"""
+    
+    #Set precedence as normal
+    teardown_module()
```
Comment (FirefoxMetzger) [inline tests/test_fits.py:88]: Could you explain why we have to tear down any module-level setup to run this test?

File: tests/test_fits.py [hunk 7]
```diff
@@ -79,5 +80,31 @@ def test_fits_reading():
...
+    
+    if IS_PYPY:
+        return  # no support for fits format :(
+
+    import numpy as np
+    from astropy.io import fits
```
Comment (FirefoxMetzger) [inline tests/test_fits.py:94]: We generally collect any imports at the top of the module. Could you refactor this and move them to the top?

File: tests/test_fits.py [hunk 8]
```diff
@@ -1,5 +1,6 @@
 """ Test fits plugin functionality.
 """
+import os
```
Comment (FirefoxMetzger) [inline tests/test_fits.py:3]:
```suggestion
```

This should no longer be needed (see comments below).

**Review by FirefoxMetzger — CHANGES_REQUESTED** [review]
Nice job! I went through your code and left some comments on how to further improve the code. 

Overall, I think it looks pretty good and we can probably merge after the coming iteration :)

**Comment by FirefoxMetzger** [comment]
Ah, the linting test fails. We strictly enforce the `black` coding style here, and also ask that a PR resolves any `flake8` issues.

To make sure the code is `black` compliant, simply run `black .` at the root of the repository.

To check for `flake8` issues run `flake8` at the root of the repository. Each problem will be listed in a new line (with reference to where it occurs). Unfortunately, they aren't fixed automatically, but it is often clear what is amiss. If the command returns without printing anything it means everything is fine :)

File: tests/test_fits.py (hunk 2 above)
Comment (astrophysaxist) [reply tests/test_fits.py:102]:
```suggestion
    hdul.writeto(tmp_path+'/test.fits')
```

Comment (astrophysaxist) [reply tests/test_fits.py:102]: Looks like `tmp_path /` is pytest syntax for temporary directories. So, I'll delete my previous suggestion and commit.

File: tests/test_fits.py (hunk 3 above)
Comment (astrophysaxist) [reply tests/test_fits.py:108]: Thanks for that suggestion! Part of the reason I ended up adding in the explicit fail was I hadn't realized the default format was set for `FITS` up earlier in the test file, so I was not testing in the right precedence scenario and was not getting a failure when I expected. Eventually saw the `teardown_module` function and was able to use it. But didn't refactor.

File: tests/test_fits.py (hunk 6 above)
Comment (astrophysaxist) [reply tests/test_fits.py:88]: Ah, I addressed this point on the `tmp_path` additions. The setup changes the precedence from normal to being default FITS format, so my test was always passing. I noticed the `teardown` function and decided to take advantage of it to set precedence levels back to normal, where the original issue occurs.

File: tests/test_fits.py (hunk 4 above)
Comment (astrophysaxist) [reply tests/test_fits.py]: Just made this change in the most recent commit.

File: tests/test_fits.py (hunk 5 above)
Comment (astrophysaxist) [reply tests/test_fits.py]: Added this decorator as well.

File: tests/test_fits.py (hunk 7 above)
Comment (astrophysaxist) [reply tests/test_fits.py]: Done!

File: tests/test_fits.py (hunk 6 above)
Comment (astrophysaxist) [reply tests/test_fits.py]: In the most recent commit I removed the `teardown_module` call and I just use the `imageio.formats.sort()` line to return the format precedence to normal for the test.

File: tests/test_fits.py (hunk 2 above)
Comment (FirefoxMetzger) [reply tests/test_fits.py]: @astrophysaxist exactly. You can find the docs here: https://docs.pytest.org/en/6.2.x/tmpdir.html#the-tmp-path-fixture

File: tests/test_fits.py (hunk 6 above)
Comment (FirefoxMetzger) [reply tests/test_fits.py]: I see, thanks for the explanation. `teardown_module` seems like a very big hammer for this. 

In another pending PR we are using the module-scope to download and generate a bunch of test images that can then easily be copied into a local folder for a test to consume. Tearing down the module may have unwanted side-effects here, such as forcing us to regenerate the files (which is expensive).

`imageio.formats.sort()` might also persist after the test has finished, since `imageio.formats` is a singleton and (I think) that libraries aren't reloaded between tests. I'll dig into this tomorrow and see if further changes are needed.

Comment (FirefoxMetzger) [reply tests/test_fits.py]: Okay, I went through the code. Both `teardown_module` and `imageio.formats.sort()` won't do the trick here, because they will both leave the environment in the wrong state after the test completes.

Instead, we need a new fixture that _temporarily_ reverts the order back to normal. It will look something like this

```python
@pytest.fixture
def normal_plugin_order():
    teardown_module()
    yield
    setup_module()
```

This will run the test using the normal plugin order and revert to the module's default state after the test is finished. It should be a fixture because that will take care of most corner cases for us, e.g. if the test crashes with an exception the environment will still be restored correctly.

We have to use that fixture by updating the test's signature `def test_fits_get_reader(normal_plugin_order, tmp_path):`.

@astrophysaxist Is this explanation sufficient to describe what has to be modified?

**Comment by FirefoxMetzger** [comment]
I took the liberty to run the code through our linter; now the linting test is green as well, meaning that all the tests are happy :)

File: tests/test_fits.py (hunk 6 above)
Comment (astrophysaxist) [reply tests/test_fits.py]: Yep that makes perfect sense @FirefoxMetzger . Happy to see pytest has this capability. Learned a lot about pytest functionality with this! Just committed those changes.

**Review by FirefoxMetzger — APPROVED** [review]

**Comment by FirefoxMetzger** [comment]
LGTM! Thanks for this contribution, and for being so responsive regarding the changes :)
